simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager.py

```diff
@@ -0,0 +1,371 @@
+            visually noisy/incorrect')
+        return None
+
+
+class RouteManager():
+    """
+    Send goals to move_base server for the specified route. Routes forever.
+
+    Loads the route from yaml.
+    Use RViz to record 2D nav goals.
+    Echo the input goal on topic /move_base_simple/goal
+    """
+
+    route_modes = {
+        'inorder': lambda goals: itertools.cycle(goals),
+        'random': lambda goals: (
+            random.choice(goals) for i in itertools.count()),
+        'dynamic': lambda goals: GoalGenerator()}
+
+    def __init__(self):
+        self.route = []
+
+        self.client = actionlib.SimpleActionClient('move_base', MoveBaseAction)
+        self.client.wait_for_server()
+
+        self.route_mode = rospy.get_param('~mode')
+        if self.route_mode not in RouteManager.route_modes:
+            rospy.logerr(
+                'Route mode %s unknown, exiting route manager',
+                self.route_mode)
+            return
+
+        poses = rospy.get_param('~poses', [])
+        if not poses and self.route_mode != 'dynamic':
+            rospy.loginfo(
+                'Route manager initialized no goals, unable to route')
+
+        self.goals = RouteManager.route_modes[self.route_mode](poses)
+        rospy.loginfo('Route manager initialized in %s mode', self.route_mode)
+
+        self.bad_goal_counter = 0
+
+    def to_move_goal(self, pose):
+        if pose is None:
+            raise ValueError('Goal position cannot be NULL')
+
+        goal = MoveBaseGoal()
+        goal.target_pose.header.stamp = rospy.Time.now()
+        goal.target_pose.header.frame_id = 'map'
+        goal.target_pose.pose.position = Point(**pose['pose']['position'])
+        goal.target_pose.pose.orientation = Quaternion(
+            **pose['pose']['orientation'])
+        return goal
+
+    def route_forever(self):
+        rate = rospy.Rate(1)
+        while not rospy.is_shutdown():
+            if self.bad_goal_counter > 10:
+                rospy.loginfo(
+                    'Stopping route manager due to too many bad goals.\
+                    Check that your occupancy map has Trinary value\
+                    representation and is not\
+                    visually noisy/incorrect'
+                    )
+                return
+            else:
+                rospy.loginfo(
+                    'Route mode is %s, getting next goal',
+                    self.route_mode)
+                try:
+                    if self.route_mode == 'dynamic':
+                        # TODO: flake8 linting standard does not allow over-riding
+                        # the built-in python method next() which is required to
+                        # make 'self.goals' a iterator.
+                        current_goal = self.to_move_goal(self.goals.get_next())
+                    else:
+                        current_goal = self.to_move_goal(next(self.goals))
```

Comment:
Are the two branches any different functionally?

Reply:
Okay, I understand now that `goals` is a different type in different `route_mode`s.